collect_app/src/main/java/org/odk/collect/android/formmanagement/FormMediaDownloader.kt

```diff
@@ -0,0 +1,63 @@
+package org.odk.collect.android.formmanagement
+
+import org.odk.collect.android.utilities.FileUtils.copyFile
+import org.odk.collect.android.utilities.FileUtils.interuptablyWriteFile
+import org.odk.collect.async.OngoingWorkListener
+import org.odk.collect.forms.Form
+import org.odk.collect.forms.FormSource
+import org.odk.collect.forms.FormSourceException
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.MediaFile
+import org.odk.collect.shared.strings.Md5.getMd5Hash
+import java.io.File
+import java.io.IOException
+
+class FormMediaDownloader(
+    private val formsRepository: FormsRepository,
+    private val formSource: FormSource
+) {
+
+    @Throws(IOException::class, FormSourceException::class, InterruptedException::class)
+    fun download(
+        formToDownload: ServerFormDetails,
+        files: List<MediaFile>,
+        tempMediaPath: String,
+        tempDir: File,
+        stateListener: OngoingWorkListener?
+    ) {
+        val tempMediaDir = File(tempMediaPath).also { it.mkdir() }
+
+        for (i in files.indices) {
+            stateListener?.progressUpdate(i + 1)
+
+            val mediaFile = files[i]
+            val tempMediaFile = File(tempMediaDir, mediaFile.filename)
+
+            val existingFile = searchForExistingMediaFile(formToDownload, mediaFile)
```

Comment:
More kotlin version would be:
```
            searchForExistingMediaFile(formToDownload, mediaFile)?.let {
                copyFile(it, tempMediaFile)
            } ?: run {
                val mediaFile = formSource.fetchMediaFile(mediaFile.downloadUrl)
                interuptablyWriteFile(mediaFile, tempMediaFile, tempDir, stateListener)
            }
```
but it's a matter of taste.

Reply:
I think you're correct that it could be more concise. How about:

```
searchForExistingMediaFile(formToDownload, mediaFile).let {
    if (it != null) {
        copyFile(it, tempMediaFile)
    } else {
        val file = formSource.fetchMediaFile(mediaFile.downloadUrl)
        interuptablyWriteFile(file, tempMediaFile, tempDir, stateListener)
    }
}
```

I find multiline "elvis" (`?:`) statements hard to follow personally.

Reply:
That's ok.